reflex/components/component.py

```diff
@@ -1235,3 +1310,273 @@ def serialize_component(comp: Component):
         The serialized component.
     """
     return str(comp)
+
+
+class StatefulComponent(BaseComponent):
+    """A component that depends on state and is rendered outside of the page component.
+
+    If a StatefulComponent is used in multiple pages, it will be rendered to a common file and
+    imported into each page that uses it.
+
+    A stateful component has a tag name that includes a hash of the code that it renders
+    to. This tag name refers to the specific component with the specific props that it
+    was created with.
+    """
+
+    # A lookup table to caching memoized component instances.
+    tag_to_stateful_component: ClassVar[Dict[str, StatefulComponent]] = {}
+
+    # Reference to the original component that was memoized into this component.
+    component: Component
+
+    # The rendered (memoized) code that will be emitted.
+    code: str
+
+    # How many times this component is referenced in the app.
+    references: int = 0
+
+    # Whether the component has already been rendered to a shared file.
+    rendered_as_shared: bool = False
+
+    @classmethod
+    def create(cls, component: Component) -> StatefulComponent | None:
+        """Create a stateful component from a component.
+
+        Args:
+            component: The component to memoize.
+
+        Returns:
+            The stateful component or None if the component should not be memoized.
+        """
+        from reflex.components.base.bare import Bare
+
+        if component.tag is None:
```

Comment:
This would be for things like `Cond` and `Foreach`... basically any subclass of `Component` that does not explicitly set a `tag` attribute will get `None`. In these cases, we assume you're doing something fancy and do no further processing.

This function returning None indicates that the component should NOT be memoized and just render it as it is.